internal/validators/oci/oci_validator.go

```diff
@@ -0,0 +1,142 @@
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/retry"
+
+	"github.com/spectrocloud-labs/validator-plugin-oci/api/v1alpha1"
+	"github.com/spectrocloud-labs/validator-plugin-oci/internal/constants"
+	vapi "github.com/spectrocloud-labs/validator/api/v1alpha1"
+	"github.com/spectrocloud-labs/validator/pkg/types"
+	vapitypes "github.com/spectrocloud-labs/validator/pkg/types"
+)
+
+type ociRule interface {
+	Name() string
+}
+
+type OciRuleService struct {
+	log logr.Logger
+}
+
+func NewOciRuleService(log logr.Logger) *OciRuleService {
+	return &OciRuleService{
+		log: log,
+	}
+}
+
+// ReconcileOciRegistryRule reconciles an OCI registry rule from the OCIValidator config
+func (s *OciRuleService) ReconcileOciRegistryRule(rule v1alpha1.OciRegistryRule) (*vapitypes.ValidationResult, error) {
+
+	// Build the default ValidationResult for this rule
+	vr := buildValidationResult(rule, constants.OciRegistry)
+
+	// Create a new registry instance
+	reg, err := remote.NewRegistry(rule.Host)
+	if err != nil {
+		s.log.V(0).Error(err, "failed to create registry client")
+		return vr, err
+	}
+
+	httpClient := retry.DefaultClient
+	// Add cert as trust
+	if rule.Cert != "" {
+		myCert, err := base64.StdEncoding.DecodeString(rule.Cert)
+		if err != nil {
+			s.log.V(0).Error(err, "failed to decode certificate", "cert", rule.Cert)
+			return vr, err
+		}
+		caCertPool, err := x509.SystemCertPool()
+		if err != nil {
+			s.log.V(0).Error(err, "failed to get system cert pool")
+			return vr, err
+		}
+		caCertPool.AppendCertsFromPEM(myCert)
+
+		// Create TLS configuration with the certificate of the server
+		tlsConfig := &tls.Config{
+			RootCAs: caCertPool,
+		}
+		httpClient.Transport = &http.Transport{TLSClientConfig: tlsConfig}
+	}
+
+	// Setup credentials if username and password are provided
+	if rule.Auth.Username != "" && rule.Auth.Password != "" {
+		reg.Client = &auth.Client{
+			Client: httpClient,
+			Cache:  auth.DefaultCache,
+			Credential: auth.StaticCredential(rule.Host, auth.Credential{
+				Username: rule.Auth.Username,
+				Password: rule.Auth.Password,
+			}),
+		}
+	}
+
+	repoPath := rule.RepositoryPath
+
+	// Get a repository path if not present
+	ctx := context.Background()
+	if repoPath == "" {
+		var repoList []string
+		err = reg.Repositories(ctx, "", func(repos []string) error {
+			for _, repo := range repos {
+				repoList = append(repoList, repo)
+			}
+			return nil
+		})
+		if err != nil {
+			s.log.V(0).Error(err, "failed to list repositories in registry", "host", rule.Host)
+			return vr, err
+		}
+
+		if len(repoList) == 0 {
+			// no repositories in registry, not possible to run any further validations
+			return vr, nil
+		}
+
+		repoPath = repoList[0]
+	}
+
+	// Get repo from regisry
+	repo, err := reg.Repository(ctx, repoPath)
```

Comment:
Is it possible that authentication would fail for certain repo, but succeed for others? I ask because in the case where no repo was specified and we arbitrarily take the first repo, maybe we should instead iterate over all the repos and try authenticating with each repo until either one succeeds or all fail?

Reply:
technically possible. i'll update accordingly